### PR TITLE
Fix Xcode 15.4 errors

### PR DESCRIPTION
**Pull request by tcobbs-bentley**
This allows imodelNative to build on macOS with the latest Xcode and a BuildStrategy of `iModelCore+iModelJsNodeAddon`.

Note: the original errors had to do with comparing a `RefCountedPtr` to `nullptr`. The only fix I could come up with for that caused comparisons to `NULL` to then fail, so I updated those places to instead use `nullptr`. There are other changes that I had to make that shouldn't have been Mac-specific. Specifically, the changes in blockcachevfsd.c and t_normals.cpp.

### CODE REVIEW

**Comment by tcobbs-bentley** [comment]
@pmconne: I'm uncomfortable with this PR. The requirement to compare to `nullptr` instead of `NULL` seems less than ideal, but I couldn't figure another solution. Without my added operators in RefCounted.h, the Mac compiler gave an error about `nullptr` and `RefCountedPtr` not being valid operands to the binary `!=` operator. And yet, I added test code that did the comparison directly, and the compiler seemed fine with it. It was just when it happened buried in a gtest macro that the compiler didn't like it.

**Comment by pmconne** [comment]
> Without my added operators in RefCounted.h, the Mac compiler gave an error about nullptr and RefCountedPtr not being valid operands to the binary != operator.

Can you show the error?

**Comment by tcobbs-bentley** [comment]
Also, I don't understand how the old code blockcachevfsd.c ever compiled anywhere. A function returning void was being used as a function pointer parameter where the parameter expected the function to return int. There's no question that the old code was broken, but I don't understand how it ever compiled.

Furthermore, the change I made in t_normals.cpp should have resulted in crashes on all architectures. It made a function call on an object that happened to be null.

**Comment by tcobbs-bentley** [comment]
> > Without my added operators in RefCounted.h, the Mac compiler gave an error about nullptr and RefCountedPtr not being valid operands to the binary != operator.
> 
> Can you show the error?

Yes, but it will take a while. I'll have to switch back to the main branch and do a clean build.

**Comment by tcobbs-bentley** [comment]
@pmconne Here is the original error output before I made the changes:

```
********************************************************************************
Error(s) for /Users/travis/src/imodelNative/src/imodel-native/iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/LinkTableRelationship_Test.cpp:
********************************************************************************
b'In file included from /Users/travis/src/imodelNative/src/imodel-native/iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/LinkTableRelationship_Test.cpp:5:'
b'In file included from /Users/travis/src/imodelNative/src/imodel-native/iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/../TestFixture/DgnDbTestFixtures.h:7:'
b'In file included from /Users/travis/src/imodelNative/src/imodel-native/iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/../TestFixture/../NonPublished/DgnHandlersTests.h:9:'
b'In file included from /Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Bentley/BeTest.h:36:'
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1454:28: error: invalid operands to binary expression ('const std::nullptr_t' and 'const BentleyM0200::RefCountedPtr<BentleyM0200::ECN::StandaloneECRelationshipEnabler>')"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1444:14: note: expanded from macro 'GTEST_IMPL_CMP_HELPER_'"
b"/Users/travis/src/imodelNative/src/imodel-native/iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/LinkTableRelationship_Test.cpp:94:5: note: in instantiation of function template specialization 'testing::internal::CmpHelperNE<std::nullptr_t, BentleyM0200::RefCountedPtr<BentleyM0200::ECN::StandaloneECRelationshipEnabler>>' requested here"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1899:31: note: expanded from macro 'ASSERT_NE'"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1881:44: note: expanded from macro 'GTEST_ASSERT_NE'"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Geom/OperatorOverload.h:174:20: note: candidate function not viable: no known conversion from 'const std::nullptr_t' to 'const DPoint3d' for 1st argument"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Geom/OperatorOverload.h:176:20: note: candidate function not viable: no known conversion from 'const std::nullptr_t' to 'const DPoint2d' for 1st argument"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1368:13: note: candidate function not viable: no known conversion from 'const std::nullptr_t' to 'faketype' for 1st argument"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1369:13: note: candidate function not viable: no known conversion from 'const std::nullptr_t' to 'faketype' for 1st argument"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Bentley/BentleyAllocator.h:136:14: note: candidate template ignored: could not match 'const BentleyAllocator<_Ty>' against 'const std::nullptr_t'"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Bentley/BentleyAllocator.h:136:14: note: candidate template ignored: could not match 'BentleyAllocator' against 'RefCountedPtr'"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Bentley/BentleyAllocator.h:142:14: note: candidate template ignored: could not match 'const BentleyAllocator<_Ty>' against 'const std::nullptr_t'"
b'In file included from /Users/travis/src/imodelNative/src/imodel-native/iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/LinkTableRelationship_Test.cpp:5:'
b'In file included from /Users/travis/src/imodelNative/src/imodel-native/iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/../TestFixture/DgnDbTestFixtures.h:7:'
b'In file included from /Users/travis/src/imodelNative/src/imodel-native/iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/../TestFixture/../NonPublished/DgnHandlersTests.h:9:'
b'In file included from /Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Bentley/BeTest.h:36:'
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1454:28: error: invalid operands to binary expression ('const std::nullptr_t' and 'const BentleyM0200::RefCountedPtr<BentleyM0200::ECN::IECRelationshipInstance>')"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1444:14: note: expanded from macro 'GTEST_IMPL_CMP_HELPER_'"
b"/Users/travis/src/imodelNative/src/imodel-native/iModelCore/iModelPlatform/Tests/DgnProject/NonPublished/LinkTableRelationship_Test.cpp:101:9: note: in instantiation of function template specialization 'testing::internal::CmpHelperNE<std::nullptr_t, BentleyM0200::RefCountedPtr<BentleyM0200::ECN::IECRelationshipInstance>>' requested here"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1899:31: note: expanded from macro 'ASSERT_NE'"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1881:44: note: expanded from macro 'GTEST_ASSERT_NE'"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Geom/OperatorOverload.h:174:20: note: candidate function not viable: no known conversion from 'const std::nullptr_t' to 'const DPoint3d' for 1st argument"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Geom/OperatorOverload.h:176:20: note: candidate function not viable: no known conversion from 'const std::nullptr_t' to 'const DPoint2d' for 1st argument"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1368:13: note: candidate function not viable: no known conversion from 'const std::nullptr_t' to 'faketype' for 1st argument"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/VendorAPI/gtest/gtest.h:1369:13: note: candidate function not viable: no known conversion from 'const std::nullptr_t' to 'faketype' for 1st argument"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Bentley/BentleyAllocator.h:136:14: note: candidate template ignored: could not match 'const BentleyAllocator<_Ty>' against 'const std::nullptr_t'"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Bentley/BentleyAllocator.h:136:14: note: candidate template ignored: could not match 'BentleyAllocator' against 'RefCountedPtr'"
b"/Users/travis/src/imodelNative/out-mac/MacOSARM64/static/BuildContexts/iModelPlatform/PublicAPI/Bentley/BentleyAllocator.h:142:14: note: candidate template ignored: could not match 'const BentleyAllocator<_Ty>' against 'const std::nullptr_t'"
```

File: iModelCore/BeSQLite/SQLite/blockcachevfsd.c [hunk 1]
```diff
@@ -810,12 +810,13 @@ static int main_copy(int argc, char **argv){
   return rc;
 }
 
-static void filesCb(void *pApp, int rc, char *z){
+static int filesCb(void *pApp, int rc, char *z){
```
Comment (nick4598) [inline iModelCore/BeSQLite/SQLite/blockcachevfsd.c:813]:
```suggestion
// BENTLEY CHANGE
static int filesCb(void *pApp, int rc, char *z){
```

File: iModelCore/BeSQLite/SQLite/blockcachevfsd.c [hunk 2]
```diff
@@ -810,12 +810,13 @@ static int main_copy(int argc, char **argv){
...
   if( rc==SQLITE_OK ){
     if( z ) printf("%s\n", z);
   }else{
     fatal_error("error (%d) - %s", rc, z);
   }
+  return rc;
```
Comment (nick4598) [inline iModelCore/BeSQLite/SQLite/blockcachevfsd.c:819]:
```suggestion
  // BENTLEY CHANGE
  return rc;
```

File: iModelCore/BeSQLite/SQLite/blockcachevfsd.c (hunk 1 above)
Comment (khanaffan) [reply iModelCore/BeSQLite/SQLite/blockcachevfsd.c:813]: We should also report this to sqlite as it there source. Updating source from cloud sqlite might accidentally override this fix again .

Comment (nick4598) [reply iModelCore/BeSQLite/SQLite/blockcachevfsd.c:813]: https://sqlite.org/cloudsqlite/forumpost/b2679c7a61

**Comment by dassaf4** [comment]
I pushed a fix for a pre-existing logic error in t_normals.cpp. `ValidatedDouble` object is quirky; it's not a smart pointer, so it's ok to access one that self-reports as "invalid". However, `PolyfaceHeaderPtr` _is_ a smart pointer, and this test was incorrectly accessing it, so that's a good fix. Thanks!

**Review by dassaf4 — APPROVED** [review]

File: iModelCore/BeSQLite/SQLite/blockcachevfsd.c (hunk 2 above)
Comment (nick4598) [inline iModelCore/BeSQLite/SQLite/blockcachevfsd.c:819]:
```suggestion
  return SQLITE_OK;
```

Comment (nick4598) [reply iModelCore/BeSQLite/SQLite/blockcachevfsd.c:819]: Apparently I missed this one change last time I updated CloudSQLite, the change on trunk returns SQLITE_OK. So going to commit that.

**Comment by tcobbs-bentley** [comment]
@pmconne Everything passed before I was forced to update from main, and now a bunch of tests are failing. Do you think this is because main became broken, or is it just intermittent test failures?

**Comment by aruniverse** [comment]
> @pmconne Everything passed before I was forced to update from main, and now a bunch of tests are failing. Do you think this is because main became broken, or is it just intermittent test failures?

If you look at the test failures, they look intermittent (issues with iTwins API rate limiting). Just re-queue them

**Review by ColinKerr — APPROVED** [review]

**Comment by tcobbs-bentley** [comment]
@Mergifyio backport release/4.8.x

**Pull request merged**